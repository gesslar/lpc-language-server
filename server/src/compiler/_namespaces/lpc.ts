export * from "../core.js";
export * from "../corePublic.js";
export * from "../types.js";
export * from "../diagnosticInformation.js";
export * from "../scanner.js";
export * from "../utilities.js";
export * from "../utilitiesPublic.js";
export * from "../parser.js";
export * from "../factory/baseNodeFactory.js";
export * from "../factory/nodeFactory.js";
export * from "../nodeTests.js";
export * from "../debug.js";
export * from "../tracing.js";
