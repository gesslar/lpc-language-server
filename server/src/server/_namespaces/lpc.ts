export * from "../../compiler/_namespaces/lpc.js";
export * from "../../services/_namespaces/lpc.js";
import * as server from "./lpc.server.js";
export { server };
