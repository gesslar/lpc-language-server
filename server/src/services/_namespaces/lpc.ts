export * from "../../compiler/_namespaces/lpc.js";
export { Symbol } from "../../compiler/_namespaces/lpc.js"; // why do I have to do this?
export * from "../types.js";
export * from "../utilities.js";
export * from "../exportInfoMap.js";
export * from "../documentRegistry.js";
export * from "../sourcemaps.js";
export * from "../services.js";

import * as FindAllReferences from "./lpc.FindAllReferences.js";
export { FindAllReferences };
import * as GoToDefinition from "./lpc.GoToDefinition.js";
export { GoToDefinition };
import * as SymbolDisplay from "./lpc.SymbolDisplay.js";
export { SymbolDisplay };