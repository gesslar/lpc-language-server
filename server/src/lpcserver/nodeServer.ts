import fs from "fs";
import net from "net";
import os from "os";
import * as lpc from "../lpc/lpc.js";
import { noop } from "../lpc/lpc.js";

export class Logger implements Logger {
    private seq = 0;
    private inGroup = false;
    private firstInGroup = true;
    private fd = -1;
    constructor(
        private readonly logFilename: string,
        private readonly traceToConsole: boolean,
        private readonly level: lpc.server.LogLevel,
    ) {
        if (this.logFilename) {
            try {
                this.fd = fs.openSync(this.logFilename, "w");
            }
            catch (_) {
                // swallow the error and keep logging disabled if file cannot be opened
            }
        }
    }
    static padStringRight(str: string, padding: string) {
        return (str + padding).slice(0, padding.length);
    }
    close() {
        if (this.fd >= 0) {
            fs.close(this.fd, noop);
        }
    }
    getLogFileName(): string | undefined {
        return this.logFilename;
    }
    perftrc(s: string) {
        this.msg(s, lpc.server.Msg.Perf);
    }
    info(s: string) {
        this.msg(s, lpc.server.Msg.Info);
    }
    err(s: string) {
        this.msg(s, lpc.server.Msg.Err);
    }
    startGroup() {
        this.inGroup = true;
        this.firstInGroup = true;
    }
    endGroup() {
        this.inGroup = false;
    }
    loggingEnabled() {
        return !!this.logFilename || this.traceToConsole;
    }
    hasLevel(level: lpc.server.LogLevel) {
        return this.loggingEnabled() && this.level >= level;
    }
    msg(s: string, type: lpc.server.Msg = lpc.server.Msg.Err) {
        if (!this.canWrite()) return;

        s = `[${lpc.server.nowString()}] ${s}\n`;
        if (!this.inGroup || this.firstInGroup) {
            const prefix = Logger.padStringRight(type + " " + this.seq.toString(), "          ");
            s = prefix + s;
        }
        this.write(s, type);
        if (!this.inGroup) {
            this.seq++;
        }
    }
    protected canWrite() {
        return this.fd >= 0 || this.traceToConsole;
    }
    protected write(s: string, _type: lpc.server.Msg) {
        if (this.fd >= 0) {
            const buf = Buffer.from(s);
            // eslint-disable-next-line no-restricted-syntax
            fs.writeSync(this.fd, buf, 0, buf.length, /*position*/ null);
        }
        if (this.traceToConsole) {
            console.debug(s);
        }
    }    
}
