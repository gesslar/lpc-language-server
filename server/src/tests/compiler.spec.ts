import * as lpc from "./_namespaces/lpc.js";
import * as lpcNode from "../lpcserver/nodeServer.js";
import * as path from "path";

/**
 * Common utilities
 */

const testPathPrefixRegExp = /(?:(file:\/{3})|\/)\.(ts|lib|src)\//g;
function removeTestPathPrefixes(text: string, retainTrailingDirectorySeparator?: boolean): string {
    return text !== undefined ? text.replace(testPathPrefixRegExp, (_, scheme) => scheme || (retainTrailingDirectorySeparator ? "/" : "")) : undefined!; // TODO: GH#18217
}

// Regex for parsing options in the format "@Alpha: Value of any sort"
const optionRegex = /^[/]{2}\s*@(\w+)\s*:\s*([^\r\n]*)/gm; // multiple matches on multiple lines

interface CompilerSettings {
    [name: string]: string;
}
function extractCompilerSettings(content: string): CompilerSettings {
    const opts: CompilerSettings = {};

    let match: RegExpExecArray | null; // eslint-disable-line no-restricted-syntax
    while ((match = optionRegex.exec(content)) !== null) { // eslint-disable-line no-restricted-syntax
        opts[match[1]] = match[2].trim();
    }

    return opts;
}


describe("Compiler", () => {

    const root = process.cwd();
    const basePath = path.join(root, "/server/src/tests/cases/compiler");
    let testFiles: string[] | undefined;

    function enumerateFiles(folder: string, regex?: RegExp, options?: { recursive: boolean; }): string[] {
        return lpc.map(lpc.listFiles(folder, regex, { recursive: (options ? options.recursive : false) }), lpc.normalizeSlashes);
    }

    it(`Should be in a node-like system`, () => {
        expect(lpc.isNodeLikeSystem()).toBeTruthy();
    });

    beforeAll(() => {
        
    });

    testFiles = enumerateFiles(basePath, /\.c$/);
    console.info(`Got ${testFiles.length} test files in ${basePath}`);

    const singleTestFile = "";//"typeInference3.c";    

    testFiles.filter(f => singleTestFile.length==0 || f.endsWith(singleTestFile)).forEach(testCaseFile => {
        const testCaseName = "compiler/" + lpc.getBaseFileName(testCaseFile);
        describe(testCaseName, () => {                  
            const fileContent = lpc.sys.readFile(testCaseFile);
            const testOptions = extractCompilerSettings(fileContent);            

            // setup options & hosts
            const compilerOptions: lpc.CompilerOptions = {
                driverType: lpc.driverTypeToLanguageVariant(testOptions["driver"]),
                diagnostics: true
            };

            const expectedErrorCount = +(testOptions["errors"] ?? "0");

            const compilerHost = lpc.createCompilerHost(compilerOptions);
            compilerHost.getDefaultLibFileName = () => lpc.combinePaths(root, lpc.getDefaultLibFileName(compilerOptions));

            const createProgramOptions: lpc.CreateProgramOptions = {
                host: compilerHost,
                rootNames: [testCaseFile],
                options: compilerOptions,                
                oldProgram: undefined,                
            };
                        
            // create program and get file we are trying to compile
            const program = lpc.createProgram(createProgramOptions);
            const file = program.getSourceFile(testCaseFile);


            let diagsText: string;
            let diags: lpc.DiagnosticWithLocation[] = [];

            it(`Properly compiles ${testCaseName}`, () => {
                expect(file).toBeDefined();
                        
                // run semantic diags
                const semanticDiags = program.getSemanticDiagnostics(file);
                
                // combine and validate
                diags = file.parseDiagnostics.concat(semanticDiags);
                diagsText = removeTestPathPrefixes(lpc.getErrorSummaryText(diags.length, lpc.getFilesInErrorForSummary(diags), "\n", compilerHost));               
            });

            it(`Reports correct number of errors for ${testCaseName}`, () => {
                expect(diags.length).toEqual(expectedErrorCount);
            });
            
            it(`Reports correct errors for ${testCaseName}`, () => {
                expect(diagsText).toMatchSnapshot(`diags-${testCaseName}`);
            });                        
        });
    });
    
});