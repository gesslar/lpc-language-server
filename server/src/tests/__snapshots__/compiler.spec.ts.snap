// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler compiler/arity1.c Reports correct errors for compiler/arity1.c: diags-compiler/arity1.c 1`] = `""`;

exports[`Compiler compiler/arity2.c Reports correct errors for compiler/arity2.c: diags-compiler/arity2.c 1`] = `
"
Found 2 errors in the same file, starting at: server/src/tests/cases/compiler/arity2.c[90m:5[0m

"
`;

exports[`Compiler compiler/assignment1.c Reports correct errors for compiler/assignment1.c: diags-compiler/assignment1.c 1`] = `""`;

exports[`Compiler compiler/assignment2.c Reports correct errors for compiler/assignment2.c: diags-compiler/assignment2.c 1`] = `
"
Found 2 errors in the same file, starting at: server/src/tests/cases/compiler/assignment2.c[90m:3[0m

"
`;

exports[`Compiler compiler/assignment3.c Reports correct errors for compiler/assignment3.c: diags-compiler/assignment3.c 1`] = `""`;

exports[`Compiler compiler/catchBlock.c Reports correct errors for compiler/catchBlock.c: diags-compiler/catchBlock.c 1`] = `""`;

exports[`Compiler compiler/closure1.c Reports correct errors for compiler/closure1.c: diags-compiler/closure1.c 1`] = `""`;

exports[`Compiler compiler/declarations1.c Reports correct errors for compiler/declarations1.c: diags-compiler/declarations1.c 1`] = `""`;

exports[`Compiler compiler/declarations2.c Reports correct errors for compiler/declarations2.c: diags-compiler/declarations2.c 1`] = `""`;

exports[`Compiler compiler/declarations3.c Reports correct errors for compiler/declarations3.c: diags-compiler/declarations3.c 1`] = `""`;

exports[`Compiler compiler/doWhile.c Reports correct errors for compiler/doWhile.c: diags-compiler/doWhile.c 1`] = `""`;

exports[`Compiler compiler/expr1.c Reports correct errors for compiler/expr1.c: diags-compiler/expr1.c 1`] = `""`;

exports[`Compiler compiler/for1.c Reports correct errors for compiler/for1.c: diags-compiler/for1.c 1`] = `""`;

exports[`Compiler compiler/for2.c Reports correct errors for compiler/for2.c: diags-compiler/for2.c 1`] = `""`;

exports[`Compiler compiler/for3.c Reports correct errors for compiler/for3.c: diags-compiler/for3.c 1`] = `""`;

exports[`Compiler compiler/functionCall1.c Reports correct errors for compiler/functionCall1.c: diags-compiler/functionCall1.c 1`] = `""`;

exports[`Compiler compiler/functionCall2.c Reports correct errors for compiler/functionCall2.c: diags-compiler/functionCall2.c 1`] = `""`;

exports[`Compiler compiler/functionCall3.c Reports correct errors for compiler/functionCall3.c: diags-compiler/functionCall3.c 1`] = `""`;

exports[`Compiler compiler/functionCall4.c Reports correct errors for compiler/functionCall4.c: diags-compiler/functionCall4.c 1`] = `""`;

exports[`Compiler compiler/functionCall5.c Reports correct errors for compiler/functionCall5.c: diags-compiler/functionCall5.c 1`] = `""`;

exports[`Compiler compiler/if1.c Reports correct errors for compiler/if1.c: diags-compiler/if1.c 1`] = `""`;

exports[`Compiler compiler/if2.c Reports correct errors for compiler/if2.c: diags-compiler/if2.c 1`] = `""`;

exports[`Compiler compiler/if3.c Reports correct errors for compiler/if3.c: diags-compiler/if3.c 1`] = `""`;

exports[`Compiler compiler/ifelse.c Reports correct errors for compiler/ifelse.c: diags-compiler/ifelse.c 1`] = `""`;

exports[`Compiler compiler/indexAccess.c Reports correct errors for compiler/indexAccess.c: diags-compiler/indexAccess.c 1`] = `""`;

exports[`Compiler compiler/jsDocParam.c Reports correct errors for compiler/jsDocParam.c: diags-compiler/jsDocParam.c 1`] = `
"
Found 1 error in server/src/tests/cases/compiler/jsDocParam.c[90m:4[0m

"
`;

exports[`Compiler compiler/lambda1.c Reports correct errors for compiler/lambda1.c: diags-compiler/lambda1.c 1`] = `""`;

exports[`Compiler compiler/macro1.c Reports correct errors for compiler/macro1.c: diags-compiler/macro1.c 1`] = `""`;

exports[`Compiler compiler/macro2.c Reports correct errors for compiler/macro2.c: diags-compiler/macro2.c 1`] = `""`;

exports[`Compiler compiler/missingFunc.c Reports correct errors for compiler/missingFunc.c: diags-compiler/missingFunc.c 1`] = `
"
Found 1 error in server/src/tests/cases/compiler/missingFunc.c[90m:3[0m

"
`;

exports[`Compiler compiler/modifiers1.c Reports correct errors for compiler/modifiers1.c: diags-compiler/modifiers1.c 1`] = `
"
Found 1 error in server/src/tests/cases/compiler/modifiers1.c[90m:3[0m

"
`;

exports[`Compiler compiler/modifiers2.c Reports correct errors for compiler/modifiers2.c: diags-compiler/modifiers2.c 1`] = `
"
Found 1 error in server/src/tests/cases/compiler/modifiers2.c[90m:2[0m

"
`;

exports[`Compiler compiler/modifiers3.c Reports correct errors for compiler/modifiers3.c: diags-compiler/modifiers3.c 1`] = `""`;

exports[`Compiler compiler/new1.c Reports correct errors for compiler/new1.c: diags-compiler/new1.c 1`] = `""`;

exports[`Compiler compiler/new2.c Reports correct errors for compiler/new2.c: diags-compiler/new2.c 1`] = `""`;

exports[`Compiler compiler/new3.c Reports correct errors for compiler/new3.c: diags-compiler/new3.c 1`] = `""`;

exports[`Compiler compiler/new4.c Reports correct errors for compiler/new4.c: diags-compiler/new4.c 1`] = `""`;

exports[`Compiler compiler/operatorBitwise.c Reports correct errors for compiler/operatorBitwise.c: diags-compiler/operatorBitwise.c 1`] = `""`;

exports[`Compiler compiler/operatorDoubleBang.c Reports correct errors for compiler/operatorDoubleBang.c: diags-compiler/operatorDoubleBang.c 1`] = `""`;

exports[`Compiler compiler/operatorLogical.c Reports correct errors for compiler/operatorLogical.c: diags-compiler/operatorLogical.c 1`] = `""`;

exports[`Compiler compiler/preproc_conditional.c Reports correct errors for compiler/preproc_conditional.c: diags-compiler/preproc_conditional.c 1`] = `""`;

exports[`Compiler compiler/preproc_conditional_macro.c Reports correct errors for compiler/preproc_conditional_macro.c: diags-compiler/preproc_conditional_macro.c 1`] = `
"
Found 1 error in server/src/tests/cases/compiler/preproc_conditional_macro.c[90m:10[0m

"
`;

exports[`Compiler compiler/struct1.c Reports correct errors for compiler/struct1.c: diags-compiler/struct1.c 1`] = `""`;

exports[`Compiler compiler/struct2.c Reports correct errors for compiler/struct2.c: diags-compiler/struct2.c 1`] = `""`;

exports[`Compiler compiler/structMemberAccess.c Reports correct errors for compiler/structMemberAccess.c: diags-compiler/structMemberAccess.c 1`] = `""`;

exports[`Compiler compiler/switch1.c Reports correct errors for compiler/switch1.c: diags-compiler/switch1.c 1`] = `""`;

exports[`Compiler compiler/switch2.c Reports correct errors for compiler/switch2.c: diags-compiler/switch2.c 1`] = `""`;

exports[`Compiler compiler/switch3.c Reports correct errors for compiler/switch3.c: diags-compiler/switch3.c 1`] = `""`;

exports[`Compiler compiler/ternary.c Reports correct errors for compiler/ternary.c: diags-compiler/ternary.c 1`] = `""`;

exports[`Compiler compiler/typeInference1.c Reports correct errors for compiler/typeInference1.c: diags-compiler/typeInference1.c 1`] = `""`;

exports[`Compiler compiler/typeInference2.c Reports correct errors for compiler/typeInference2.c: diags-compiler/typeInference2.c 1`] = `""`;

exports[`Compiler compiler/typeInference3.c Reports correct errors for compiler/typeInference3.c: diags-compiler/typeInference3.c 1`] = `
"
Found 1 error in server/src/tests/cases/compiler/typeInference3.c[90m:4[0m

"
`;

exports[`Compiler compiler/varBeforeDecl.c Reports correct errors for compiler/varBeforeDecl.c: diags-compiler/varBeforeDecl.c 1`] = `
"
Found 1 error in server/src/tests/cases/compiler/varBeforeDecl.c[90m:3[0m

"
`;

exports[`Compiler compiler/varBeforeDecl2.c Reports correct errors for compiler/varBeforeDecl2.c: diags-compiler/varBeforeDecl2.c 1`] = `
"
Found 1 error in server/src/tests/cases/compiler/varBeforeDecl2.c[90m:2[0m

"
`;

exports[`Compiler compiler/varBeforeDecl3.c Reports correct errors for compiler/varBeforeDecl3.c: diags-compiler/varBeforeDecl3.c 1`] = `""`;

exports[`Compiler compiler/while1.c Reports correct errors for compiler/while1.c: diags-compiler/while1.c 1`] = `""`;
